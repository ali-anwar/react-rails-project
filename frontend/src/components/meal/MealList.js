import React from "react";
import PropTypes from 'prop-types';
import { BootstrapTable, TableHeaderColumn } from "react-bootstrap-table";
import { formatTimeDisplay } from '../../lib/timeConverter';

const DEFAULT_ROW_STYLE = {
  backgroundColor: ""
};

const getCaret = direction => {
  if (direction === "asc") {
    return (
      <span>
        {" "}
        <i className="fa fa-sort-asc" aria-hidden="true" />
      </span>
    );
  }

  if (direction === "desc") {
    return (
      <span>
        {" "}
        <i className="fa fa-sort-desc" aria-hidden="true" />
      </span>
    );
  }

  return (
    <span>
      {" "}
      <i className="fa fa-sort" aria-hidden="true" />
    </span>
  );
};

const buttonFormatter = (cell, row, props) => {
  return (
    <div>
      <button
        className="btn btn-warning mr-2"
        onClick={() => props.handleEditButton(row.id)}
      >
        <i className="fa fa-pencil" aria-hidden="true" />
      </button>

      <button
        className="btn btn-danger"
        onClick={() => props.handleDeleteButton(row.id)}
      >
        <i className="fa fa-trash" aria-hidden="true" />
      </button>
    </div>
  );
};

const timeFormatter = (cell, row) => {
  return `${formatTimeDisplay(cell)}`
}

class MealList extends React.Component {
  constructor(props) {
    super(props);

    this.options = {
      sortIndicator: true,
      noDataText: "No data"
    };

    this.selectRowProp = {
      mode: "radio",
      onSelect: props.handleRowSelect,
      clickToSelect: true,
      hideSelectColumn: true
    };
  }

  rowStyleFormat = (row, rowIdx) => {
    if (!row || row.noCaloriesLimit) {
      return DEFAULT_ROW_STYLE;
    }
    return {
      backgroundColor: row.exceedsCaloriesLimit ? "#ff8c8c" : "#63de76"
    };
  };

  render() {
    let userNameRow = null;

    if(this.props.isCurrentAdminUser) {
      userNameRow = (
        <TableHeaderColumn
          dataField="userName"
          dataSort={true}
          caretRender={getCaret}
          columnTitle
        >
          User Name
        </TableHeaderColumn>
      );
    }
    return (
      <BootstrapTable
        data={this.props.meals}
        selectRow={this.selectRowProp}
        options={this.options}
        bordered={false}
        trStyle={
          this.props.isCurrentAdminUser
            ? DEFAULT_ROW_STYLE
            : this.rowStyleFormat
        }
        className="custom-table"
        striped
        hover
        condensed
      >
        <TableHeaderColumn dataField="id" isKey hidden>
          Id
        </TableHeaderColumn>

        <TableHeaderColumn
          dataField="text"
          dataSort={true}
          caretRender={getCaret}
          columnTitle
        >
          Text
        </TableHeaderColumn>

        <TableHeaderColumn
          dataField="no_of_calories"
          dataSort={true}
          caretRender={getCaret}
          columnTitle
        >
          Calories
        </TableHeaderColumn>

        <TableHeaderColumn
          dataField="date"
          dataSort={true}
          caretRender={getCaret}
          columnTitle
        >
          Date
        </TableHeaderColumn>

        <TableHeaderColumn
          dataField="time"
          dataSort={true}
          dataFormat={timeFormatter}
          caretRender={getCaret}
          columnTitle
        >
          Time
        </TableHeaderColumn>

        {userNameRow}

        <TableHeaderColumn
          dataField="button"
          dataFormat={buttonFormatter}
          formatExtraData={this.props}
        >
          Actions
        </TableHeaderColumn>
      </BootstrapTable>
    );
  }
}

MealList.propTypes = {
  meals: PropTypes.array.isRequired,
  handleRowSelect: PropTypes.func.isRequired,
  isCurrentAdminUser: PropTypes.bool.isRequired,
  handleDeleteButton: PropTypes.func.isRequired,
  handleEditButton: PropTypes.func.isRequired
};

export default MealList;
