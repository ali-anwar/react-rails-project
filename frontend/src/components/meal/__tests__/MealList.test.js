import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import MealList from '../MealList';


describe('MealList.test.js', () => {

  it('renders without crashing', () => {
    const props = {
      meals: [{
        id: "react-flux-building-applications",
        title: "Building Applications in React and Flux",
        calories: 100,
        description: "Very Tasty"
      }],
      handleRowSelect: jest.fn(),
      isCurrentAdminUser: false,
      handleDeleteButton: jest.fn(),
      handleEditButton: jest.fn(),
    };
    
    const wrapper = shallow(<MealList {...props}/>);

    expect(wrapper).toHaveLength(1);
    expect(wrapper.find('BootstrapTable')).toHaveLength(1);

    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });
});
