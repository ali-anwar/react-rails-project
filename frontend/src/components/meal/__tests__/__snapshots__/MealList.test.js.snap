exports[`MealList.test.js renders without crashing 1`] = `
<BootstrapTable
  bodyContainerClass={null}
  bordered={false}
  cellEdit={
    Object {
      "afterSaveCell": undefined,
      "beforeSaveCell": undefined,
      "blurToSave": false,
      "mode": "none",
      "nonEditableRows": undefined,
    }
  }
  className="custom-table"
  columnFilter={false}
  condensed={true}
  containerClass={null}
  csvFileName="spreadsheet.csv"
  data={
    Array [
      Object {
        "calories": 100,
        "description": "Very Tasty",
        "id": "react-flux-building-applications",
        "title": "Building Applications in React and Flux",
      },
    ]
  }
  deleteRow={false}
  expandColumnOptions={
    Object {
      "expandColumnBeforeSelectColumn": true,
      "expandColumnComponent": undefined,
      "expandColumnVisible": false,
    }
  }
  exportCSV={false}
  fetchInfo={
    Object {
      "dataTotalSize": 0,
    }
  }
  headerContainerClass={null}
  height="100%"
  hover={true}
  ignoreSinglePage={false}
  insertRow={false}
  keyBoardNav={false}
  multiColumnSearch={false}
  multiColumnSort={1}
  options={
    Object {
      "noDataText": "No data",
      "sortIndicator": true,
    }
  }
  pagination={false}
  printable={false}
  replace={false}
  search={false}
  selectRow={
    Object {
      "clickToSelect": true,
      "hideSelectColumn": true,
      "mode": "radio",
      "onSelect": [Function],
    }
  }
  striped={true}
  tableBodyClass={null}
  tableContainerClass={null}
  tableHeaderClass={null}
  trClassName=""
  trStyle={[Function]}
  withoutTabIndex={false}>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    dataAlign="left"
    dataField="id"
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={true}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={true}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}>
    Id
  </TableHeaderColumn>
  <TableHeaderColumn
    caretRender={[Function]}
    className=""
    columnClassName=""
    columnTitle={true}
    dataAlign="left"
    dataField="text"
    dataSort={true}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}>
    Text
  </TableHeaderColumn>
  <TableHeaderColumn
    caretRender={[Function]}
    className=""
    columnClassName=""
    columnTitle={true}
    dataAlign="left"
    dataField="no_of_calories"
    dataSort={true}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}>
    Calories
  </TableHeaderColumn>
  <TableHeaderColumn
    caretRender={[Function]}
    className=""
    columnClassName=""
    columnTitle={true}
    dataAlign="left"
    dataField="date"
    dataSort={true}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}>
    Date
  </TableHeaderColumn>
  <TableHeaderColumn
    caretRender={[Function]}
    className=""
    columnClassName=""
    columnTitle={true}
    dataAlign="left"
    dataField="time"
    dataFormat={[Function]}
    dataSort={true}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}>
    Time
  </TableHeaderColumn>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    dataAlign="left"
    dataField="button"
    dataFormat={[Function]}
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    formatExtraData={
      Object {
        "handleDeleteButton": [Function],
        "handleEditButton": [Function],
        "handleRowSelect": [Function],
        "isCurrentAdminUser": false,
        "meals": Array [
          Object {
            "calories": 100,
            "description": "Very Tasty",
            "id": "react-flux-building-applications",
            "title": "Building Applications in React and Flux",
          },
        ],
      }
    }
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}>
    Actions
  </TableHeaderColumn>
</BootstrapTable>
`;
